### PR TITLE
Do not report errors when inference is partially blocked

### PR DESCRIPTION
**Pull request by Andarist**
fixes https://github.com/microsoft/TypeScript/issues/50818
sort of supersedes https://github.com/microsoft/TypeScript/pull/52393

cc @andrewbranch @weswigham

### CODE REVIEW

File: src/compiler/checker.ts
Comment (Andarist) [inline src/compiler/checker.ts:32607]: alternatively, we could check `checkMode` but `ContextFlags.Completions` isn't currently passed down as any kind of `checkMode` (whereas `CheckMode.IsForStringLiteralArgumentCompletions` already exists). So this would require some additional plumbing and I'm not sure if that's desired

File: tests/cases/fourslash/noErrorsAfterCompletionsRequestWithinGenericFunction1.ts
Comment (Andarist) [inline tests/cases/fourslash/noErrorsAfterCompletionsRequestWithinGenericFunction1.ts:9]: this was reporting excess property check - I still think that perhaps it **shouldn't** cause applicability error in this case when requesting completions but that's a separate problem

File: tests/cases/fourslash/noErrorsAfterCompletionsRequestWithinGenericFunction3.ts
Comment (Andarist) [inline tests/cases/fourslash/noErrorsAfterCompletionsRequestWithinGenericFunction3.ts:25]: This was an interesting case because the whole argument got blocked and thus `A` and `B` were replaced with their constraints and thus typing the argument `Funcs<unknown, Record<string, unknown>>`. And then the argument became not assignable to this target type - see the playground [here](https://www.typescriptlang.org/play?ssl=22&ssc=15&pln=22&pc=54#code/C4TwDgpgBAYgrgOwMYGcA8BBANFAQlCAD2AgQBMUoAlCJAewCcy0VgGBLBAcx0QGsEdAO4IAfKKgBeKAG8AUFCgBtANJROUPhBB0AZngC6ALlkLFUXQhMAKAIYnsUAEYncASikSAbnXZkA3GaKwAAWnFyuqgaBigC+gfFycmS0ADa2DNC6iEjA7HQIFnR0mDj4RCTklDT0TCxs4bwIAsJiotZmligm8Mjojriicm4mSgPRSaCQUADKdHAMSNDS8opOGSar5pY29lCsHNw4LrLOG-sN3P5nAF4mB+FQsR6S3r4BQVCh4feXXDFPAHrO6mcwWKxQOy-Q48ZybM4MaHha7ApHcJ4vN5+AHBMLcNH-MyJRJTaAAFQyXAgwCksBy6H4ghEOBqjGYDyOUEZrXEgTkpKgAGEQrQ+LS5gslgRiKQKFAKQwqTSAPxQACMUBMAAZAgB6XXmAB6yrkQA)

File: src/compiler/checker.ts
Comment (jakebailey) [inline src/compiler/checker.ts:1830]: Is this reentrant such that you should save the value and restore it, rather than unconditionally setting false?

Comment (Andarist) [reply src/compiler/checker.ts:1830]: I intentionally ignored this to keep the implementation simpler as I dont think it is

**Comment by DanielRosenwasser** [comment]
@typescript-bot pack this
@typescript-bot user test tsserver
@typescript-bot test tsserver top100

**Comment by andrewbranch** [comment]
This looks good to me, but will wait until our extended tests work again to approve 🥲

**Comment by jakebailey** [comment]
The test stuff is working again, but this PR has a conflict now so I don't know if they'd actually give the right results.

**Comment by Andarist** [comment]
@jakebailey conflicts resolved

**Comment by jakebailey** [comment]
@typescript-bot test this
@typescript-bot test top100
@typescript-bot user test this
@typescript-bot user test tsserver
@typescript-bot test tsserver top100
@typescript-bot run dt
@typescript-bot perf test this
@typescript-bot pack this

**Review by andrewbranch — APPROVED** [review]

**Comment by jakebailey** [comment]
Just to check my understanding, this silences the errors, but is it possible that information related to those errors persists in some cache?

**Comment by Andarist** [comment]
This silences error reporting but before `diagnostics.add` is called - so no, nothing should be put into the cache. If anything would be cached during this call then it would be a bug - those calls are sensitive to supplied "work in progress" input and thus caching has to be severely limited during those calls.

**Comment by Andarist** [comment]
Ye, if I understand correctly the relationship cache is stable~ since it just caches the results of comparing A and B but that's totally unrelated to where those A and B are coming from.

**Pull request merged**
